src/fsharp/service/IncrementalBuild.fs

```diff
@@ -1506,188 +585,282 @@ type IncrementalBuilder(tcGlobals, frameworkTcImports, nonFrameworkAssemblyInput
+
+    let invalidateBuild () =
+        initialTcAccCache <- None
+        invalidateSlot 0
+
+    let getSlot filename =
+        typeCheckCache 
+        |> Array.findIndex (fun x ->
+            let f2 = x.FileName
+            String.Compare(filename, f2, StringComparison.CurrentCultureIgnoreCase)=0
+            || String.Compare(FileSystem.GetFullPathShim filename, FileSystem.GetFullPathShim f2, StringComparison.CurrentCultureIgnoreCase)=0)
+
+    let getParseResults ctok slot =
+        checkSlot slot
+
+        match typeCheckCache.[slot] with
+        | Parsed(_, parsedInfo) -> parsedInfo
+        | NotParsed sourceFile ->
+            let parsedInfo = parse ctok sourceFile slot
+            typeCheckCache.[slot] <- Parsed(sourceFile, parse ctok sourceFile slot)
+            parsedInfo
+        | Checked(sourceFile, _, weakInfo) ->
+            match weakInfo.TryGetTarget() with
+            | true, parsedInfo -> parsedInfo
+            | _ -> 
+                let parsedInfo = parse ctok sourceFile slot
+                weakInfo.SetTarget parsedInfo
+                parsedInfo
+
+    let sourceFileTimeStamps =
+        sourceFiles
+        |> Array.ofList
+        |> Array.map (fun (_, nm, _) -> (nm, DateTime()))
+
+    /// Checks source file timestamps.
+    /// Source files can be invalidated by this call, which will cause re-type-checking.
+    let checkSourceFileTimeStamps (cache: TimeStampCache) =
+        sourceFileTimeStamps
+        |> Array.iteri (fun slot (filename, currentTimeStamp) ->
+            let newTimeStamp = cache.GetFileTimeStamp filename
+            if newTimeStamp <> currentTimeStamp then
+                invalidateSlot slot
+                sourceFileTimeStamps.[slot] <- (filename, newTimeStamp))
+
+    let referenceAssemblyTimeStamps =
+        nonFrameworkAssemblyInputs
+        |> Array.ofList
+        |> Array.map (fun (_, getTimeStamp) -> (getTimeStamp, DateTime()))
+
+    /// Checks reference assembly timestamps.
+    /// Build can be invalidated by this call.
+    let checkReferenceAssemblyTimeStamps ctok (cache: TimeStampCache) =
+        referenceAssemblyTimeStamps
+        |> Array.iteri (fun i (getTimeStamp, currentTimeStamp) ->
+            let newTimeStamp = getTimeStamp cache ctok
+            if newTimeStamp <> currentTimeStamp then
+                invalidateBuild ()
+                referenceAssemblyTimeStamps.[i] <- (getTimeStamp, newTimeStamp))
+
+    /// Checks source files and reference assembly timestamps.
+    /// Source files and/or build can be invalidated by this call.
+    let checkTimeStamps ctok cache =
+        checkSourceFileTimeStamps cache
+        checkReferenceAssemblyTimeStamps ctok cache
+
+    let tryCreatePartialCheckResultsBeforeSlot slot =
+        match slot, initialTcAccCache with
+        | _, None -> None
+        | 0, Some initialTcAcc ->
+            Some(PartialCheckResults.Create(initialTcAcc, defaultTimeStamp))
+        | _ -> 
```

Comment:
What if `slot` is very small or large?

Reply:
`checkSlot` will throw if it's out of range.